rust/src/dataset/write.rs

```diff
@@ -53,3 +73,300 @@ impl Default for WriteParams {
+                if rows_to_take == rows_remaining_in_batch {
+                    // We're taking the whole batch, so we can just move it
+                    let batch = if self.i == 0 {
+                        batch
+                    } else {
+                        // We are taking the remainder of the batch, so we need to slice it
+                        batch.slice(self.i, rows_to_take)
+                    };
+                    batches.push(batch);
+                    self.i = 0;
+                } else {
+                    // We're taking a slice of the batch, so we need to copy it
+                    batches.push(batch.slice(self.i, rows_to_take));
+                    // And then we need to push the remainder back onto the front of the queue
+                    self.i += rows_to_take;
+                    self.buffered.push_front(batch);
+                }
+
+                rows_collected += rows_to_take;
+            } else {
+                break;
+            }
+        }
+
+        if batches.is_empty() {
+            None
+        } else {
+            Some(Ok(batches))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use arrow::array::Int32Array;
+    use arrow_schema::DataType;
+    use arrow_schema::Schema as ArrowSchema;
+
+    #[tokio::test]
+    async fn test_chunking_large_batches() {
+        // Create a stream of 3 batches of 10 rows
+        let schema = Arc::new(ArrowSchema::new(vec![arrow::datatypes::Field::new(
+            "a",
+            DataType::Int32,
+            false,
+        )]));
+        let batch =
+            RecordBatch::try_new(schema.clone(), vec![Arc::new(Int32Array::from_iter(0..28))])
+                .unwrap();
+        let batches: Vec<RecordBatch> =
+            vec![batch.slice(0, 10), batch.slice(10, 10), batch.slice(20, 8)];
+        let stream = RecordBatchStreamAdapter::new(
+            schema.clone(),
+            futures::stream::iter(batches.into_iter().map(Ok::<_, DataFusionError>)),
+        );
+
+        // Chunk into a stream of 3 row batches
+        let chunks: Vec<Vec<RecordBatch>> = chunk_stream(Box::pin(stream), 3)
+            .try_collect()
+            .await
+            .unwrap();
+
+        assert_eq!(chunks.len(), 10);
+        assert_eq!(chunks[0].len(), 1);
+
+        for (i, chunk) in chunks.iter().enumerate() {
+            let num_rows = chunk.iter().map(|batch| batch.num_rows()).sum::<usize>();
+            if i < chunks.len() - 1 {
+                assert_eq!(num_rows, 3);
+            } else {
+                // Last chunk is shorter
+                assert_eq!(num_rows, 1);
+            }
+        }
+
+        // The fourth chunk is split along the boundary between the original first
+        // two batches.
+        assert_eq!(chunks[3].len(), 2);
+        assert_eq!(chunks[3][0].num_rows(), 1);
+        assert_eq!(chunks[3][1].num_rows(), 2);
+    }
+
+    #[tokio::test]
+    async fn test_chunking_small_batches() {
```

Comment:
Could you add one more test, that it has a one single batch with , say, 1000 rows, and write with `max_rows_per_group=10`, `max_rows_perf_file=100`.

Reply:
Added such a test to `dataset.rs` 👍